File: auth-client/python/src/mcp_auth_client/verifier.py
import asyncio
import ipaddress
import json
import time
from dataclasses import dataclass
from typing import Any, cast
from urllib.parse import urlparse

import httpx
import jwt
from jwt.algorithms import RSAAlgorithm

MAX_JWKS_BYTES = 1 << 20


class TokenVerificationError(ValueError):
    """Raised for any invalid inbound MCP access token."""


@dataclass(frozen=True, slots=True)
class TokenClaims:
    subject: str
    issuer: str
    audience: str | list[str]
    scopes: frozenset[str]
    claims: dict[str, Any]


class JWTVerifier:
    def __init__(
        self,
        jwks_uri: str,
        issuer: str,
        audience: str,
        required_scopes: set[str] | frozenset[str] = frozenset(),
        algorithms: tuple[str, ...] = ("RS256",),
        http_client: httpx.AsyncClient | None = None,
        jwks_ttl: float = 300.0,
        ssrf_safe: bool = True,
        clock_skew: float = 60.0,
        http_timeout: float = 10.0,
        jwks_min_fetch: float = 30.0,
        unknown_key_ttl: float = 30.0,
    ) -> None:
        if not algorithms:
            raise ValueError("at least one JWT algorithm must be allowlisted")
        if any(algorithm not in {"RS256", "RS384", "RS512"} for algorithm in algorithms):
            raise ValueError("only RSA SHA-2 algorithms are supported by the default verifier")
        if clock_skew < 0:
            raise ValueError("clock_skew must be non-negative")
        if http_timeout <= 0:
            raise ValueError("http_timeout must be positive")
        if jwks_min_fetch < 0 or unknown_key_ttl < 0:
            raise ValueError("JWKS cache intervals must be non-negative")
        self.jwks_uri, self.issuer, self.audience = jwks_uri, issuer, audience
        self.required_scopes = frozenset(required_scopes)
        self.algorithms = algorithms
        self._http_client = http_client
        self._jwks_ttl = jwks_ttl
        self.ssrf_safe = ssrf_safe
        self._clock_skew = clock_skew
        self._http_timeout = http_timeout
        self._jwks_min_fetch = jwks_min_fetch
        self._unknown_key_ttl = unknown_key_ttl
        self._jwks: dict[str, Any] = {}
        self._jwks_loaded_at = 0.0
        self._last_refresh = 0.0
        self._unknown_kids: dict[str, float] = {}
        self._jwks_lock = asyncio.Lock()

    @classmethod
    def from_jwks(cls, jwks: dict[str, object], **kwargs: object) -> "JWTVerifier":
        verifier = cls(jwks_uri="in-memory://jwks", **kwargs)  # type: ignore[arg-type]
        verifier._set_jwks(jwks)
        return verifier

    async def verify(self, token: str) -> TokenClaims:
        try:
            header = jwt.get_unverified_header(token)
            algorithm = header.get("alg")
            kid = header.get("kid")
            typ = header.get("typ")
            if not isinstance(algorithm, str) or algorithm not in self.algorithms:
                raise TokenVerificationError("JWT algorithm is not allowlisted")
            # A present typ must be a string. Guarding on isinstance let a
            # number, list, or object through unchecked, because none of them
            # is one of the two allowed values.
            if typ is not None and (not isinstance(typ, str) or typ not in {"JWT", "at+jwt"}):
                raise TokenVerificationError("JWT type is not allowed")
            if not isinstance(kid, str):
                raise TokenVerificationError("JWT kid is missing")
            jwk = await self._ensure_key(kid)
            if jwk is None:
                raise TokenVerificationError("JWT signing key was not found")
            key = RSAAlgorithm.from_jwk(json.dumps(jwk))
            claims = jwt.decode(
                token,
                key=cast(Any, key),
                algorithms=list(self.algorithms),
                issuer=self.issuer,
                audience=self.audience,
                leeway=self._clock_skew,
                options={"require": ["exp", "iss", "sub", "aud"]},
            )
            scopes = frozenset(str(claims.get("scope", "")).split())
            if not self.required_scopes.issubset(scopes):
                raise TokenVerificationError("JWT does not contain all required scopes")
            audience = claims["aud"]
            if not isinstance(audience, (str, list)):
                raise TokenVerificationError("JWT audience has an invalid type")
            return TokenClaims(str(claims["sub"]), str(claims["iss"]), audience, scopes, claims)
        except TokenVerificationError:
            raise
        except (jwt.PyJWTError, TypeError, ValueError, KeyError) as exc:
            raise TokenVerificationError("invalid JWT") from exc

    async def _ensure_key(self, kid: str) -> dict[str, Any] | None:
        if self.jwks_uri.startswith("in-memory://"):
            if not self._jwks:
                raise TokenVerificationError("in-memory JWKS was not configured")
            return cast(dict[str, Any] | None, self._jwks.get(kid))
        now = time.time()
        key = self._jwks.get(kid)
        fresh = bool(self._jwks) and now - self._jwks_loaded_at < self._jwks_ttl
        if isinstance(key, dict) and fresh:
            return key
        if self._recent_miss(kid, now):
            return None
        async with self._jwks_lock:
            now = time.time()
            key = self._jwks.get(kid)
            fresh = bool(self._jwks) and now - self._jwks_loaded_at < self._jwks_ttl
            if isinstance(key, dict) and fresh:
                return key
            if self._last_refresh and now - self._last_refresh < self._jwks_min_fetch:
                # Throttled. A key already in the cache still verifies the
                # signature, so serve it rather than rejecting a valid token:
                # returning None here reported "signing key was not found" for
                # a kid sitting in self._jwks whenever jwks_ttl was shorter
                # than jwks_min_fetch. Only a genuinely absent kid fails.
                return key if isinstance(key, dict) else None
            await self._fetch_jwks()
            found = self._jwks.get(kid)
            if isinstance(found, dict):
                self._unknown_kids.pop(kid, None)
                return found
            self._unknown_kids[kid] = time.time()
            return None

    def _recent_miss(self, kid: str, now: float) -> bool:
        missed = self._unknown_kids.get(kid)
        if missed is None:
            return False
        return (
            now - missed < self._unknown_key_ttl and now - self._last_refresh < self._jwks_min_fetch
        )

    async def _fetch_jwks(self) -> None:
        self._validate_jwks_uri()
        owns_client = self._http_client is None
        timeout = httpx.Timeout(self._http_timeout)
        client = self._http_client or httpx.AsyncClient(timeout=timeout)
        try:
            response = await client.get(
                self.jwks_uri,
                headers={"Accept": "application/json"},
                timeout=timeout,
            )
            response.raise_for_status()
            body = await self._read_limited(response)
            self._set_jwks(json.loads(body))
        except TokenVerificationError:
            raise
        except (httpx.HTTPError, ValueError, TypeError) as exc:
            raise TokenVerificationError("unable to load JWKS") from exc
        finally:
            if owns_client:
                await client.aclose()

    async def _read_limited(self, response: httpx.Response) -> bytes:
        declared = response.headers.get("Content-Length")
        if declared is not None:
            try:
                length = int(declared)
            except ValueError as exc:
                raise TokenVerificationError("unable to load JWKS") from exc
            if length > MAX_JWKS_BYTES:
                raise TokenVerificationError("JWKS response is too large")
        chunks: list[bytes] = []
        size = 0
        async for chunk in response.aiter_bytes():
            size += len(chunk)
            if size > MAX_JWKS_BYTES:
                raise TokenVerificationError("JWKS response is too large")
            chunks.append(chunk)
        return b"".join(chunks)

    def _validate_jwks_uri(self) -> None:
        parsed = urlparse(self.jwks_uri)
        if parsed.scheme not in {"https", "http"} or not parsed.hostname:
            raise TokenVerificationError("JWKS URI must be an absolute HTTP(S) URL")
        if not self.ssrf_safe:
            return
        if parsed.username or parsed.password:
            raise TokenVerificationError("JWKS URI must not contain credentials")
        if parsed.scheme == "http" and parsed.hostname not in {"localhost", "127.0.0.1", "::1"}:
            raise TokenVerificationError("SSRF-safe JWKS loading requires HTTPS")
        try:
            address = ipaddress.ip_address(parsed.hostname)
        except ValueError:
            return
        if not address.is_loopback:
            raise TokenVerificationError("SSRF-safe JWKS loading rejects non-loopback IP addresses")

    def _set_jwks(self, data: object) -> None:
        if not isinstance(data, dict) or not isinstance(data.get("keys"), list):
            raise TokenVerificationError("JWKS response is invalid")
        self._jwks = {
            str(item["kid"]): item
            for item in data["keys"]
            if isinstance(item, dict) and isinstance(item.get("kid"), str)
        }
        now = time.time()
        self._jwks_loaded_at = now
        self._last_refresh = now

File: auth-client/typescript/src/verifier.ts
import { createPublicKey, verify as verifySignature, type JsonWebKey } from "node:crypto";
import { isIP } from "node:net";

import { DEFAULT_TIMEOUT_MS, requestTimeout } from "./timeout.js";

const MAX_JWKS_BYTES = 1024 * 1024;
const DEFAULT_JWKS_MIN_FETCH_MS = 30_000;
const DEFAULT_UNKNOWN_KEY_TTL_MS = 30_000;

export class TokenVerificationError extends Error {
  constructor(message = "invalid MCP access token", options?: ErrorOptions) {
    super(message, options);
    this.name = "TokenVerificationError";
  }
}

export interface TokenClaims {
  subject: string;
  issuer: string;
  audience: string[];
  scopes: ReadonlySet<string>;
  claims: Readonly<Record<string, unknown>>;
}

export interface JWTVerifierOptions {
  jwksUri: string;
  issuer: string;
  audience: string;
  requiredScopes?: Iterable<string>;
  fetch?: typeof globalThis.fetch;
  jwksTtlMs?: number;
  clockSkewSeconds?: number;
  ssrfSafe?: boolean;
  timeoutMs?: number;
  jwksMinFetchMs?: number;
  unknownKeyTtlMs?: number;
}

interface JWKSDocument { keys: JsonWebKey[] }

function decodePart(value: string): Record<string, unknown> {
  try {
    const decoded: unknown = JSON.parse(Buffer.from(value, "base64url").toString("utf8"));
    if (decoded === null || typeof decoded !== "object" || Array.isArray(decoded)) throw new Error();
    return decoded as Record<string, unknown>;
  } catch (error) {
    throw new TokenVerificationError("JWT contains invalid JSON", { cause: error });
  }
}

function audiences(value: unknown): string[] {
  if (typeof value === "string") return [value];
  if (Array.isArray(value) && value.every((item) => typeof item === "string")) return value;
  return [];
}

export class JWTVerifier {
  readonly jwksUri: string;
  readonly issuer: string;
  readonly audience: string;
  readonly requiredScopes: ReadonlySet<string>;
  private readonly fetcher: typeof globalThis.fetch;
  private readonly jwksTtlMs: number;
  private readonly clockSkewSeconds: number;
  private readonly ssrfSafe: boolean;
  private readonly timeoutMs: number;
  private readonly jwksMinFetchMs: number;
  private readonly unknownKeyTtlMs: number;
  private keys = new Map<string, JsonWebKey>();
  private loadedAt = 0;
  private lastRefresh = 0;
  private readonly unknownKids = new Map<string, number>();
  private loading: Promise<void> | undefined;

  constructor(options: JWTVerifierOptions) {
    this.jwksUri = options.jwksUri;
    this.issuer = options.issuer;
    this.audience = options.audience;
    this.requiredScopes = new Set(options.requiredScopes ?? []);
    this.fetcher = options.fetch ?? globalThis.fetch;
    this.jwksTtlMs = options.jwksTtlMs ?? 300_000;
    this.clockSkewSeconds = options.clockSkewSeconds ?? 60;
    this.ssrfSafe = options.ssrfSafe ?? true;
    this.timeoutMs = options.timeoutMs ?? DEFAULT_TIMEOUT_MS;
    this.jwksMinFetchMs = options.jwksMinFetchMs ?? DEFAULT_JWKS_MIN_FETCH_MS;
    this.unknownKeyTtlMs = options.unknownKeyTtlMs ?? DEFAULT_UNKNOWN_KEY_TTL_MS;
  }

  async verify(token: string, signal?: AbortSignal): Promise<TokenClaims> {
    const parts = token.split(".");
    if (parts.length !== 3) throw new TokenVerificationError();
    const [encodedHeader, encodedPayload, encodedSignature] = parts as [string, string, string];
    const header = decodePart(encodedHeader);
    const claims = decodePart(encodedPayload);
    if (header.alg !== "RS256") throw new TokenVerificationError("JWT algorithm is not allowed");
    if (header.typ !== undefined && header.typ !== "JWT" && header.typ !== "at+jwt") {
      throw new TokenVerificationError("JWT type is not allowed");
    }
    if (typeof header.kid !== "string" || !header.kid) throw new TokenVerificationError("JWT kid is missing");

    const jwk = await this.signingKey(header.kid, signal);
    if (!jwk) throw new TokenVerificationError("JWT signing key was not found");
    let validSignature = false;
    try {
      validSignature = verifySignature(
        "RSA-SHA256",
        Buffer.from(`${encodedHeader}.${encodedPayload}`),
        createPublicKey({ key: jwk, format: "jwk" }),
        Buffer.from(encodedSignature, "base64url"),
      );
    } catch (error) {
      throw new TokenVerificationError("JWT signing key is invalid", { cause: error });
    }
    if (!validSignature) throw new TokenVerificationError("JWT signature is invalid");

    const now = Date.now() / 1000;
    if (claims.iss !== this.issuer) throw new TokenVerificationError("JWT issuer is invalid");
    if (typeof claims.exp !== "number" || now > claims.exp + this.clockSkewSeconds) {
      throw new TokenVerificationError("JWT is expired");
    }
    if (typeof claims.nbf === "number" && now + this.clockSkewSeconds < claims.nbf) {
      throw new TokenVerificationError("JWT is not active");
    }
    const tokenAudiences = audiences(claims.aud);
    if (!tokenAudiences.includes(this.audience)) throw new TokenVerificationError("JWT audience is invalid");
    if (typeof claims.sub !== "string" || !claims.sub) throw new TokenVerificationError("JWT subject is missing");
    const scopes = new Set(typeof claims.scope === "string" ? claims.scope.split(/\s+/).filter(Boolean) : []);
    for (const scope of this.requiredScopes) {
      if (!scopes.has(scope)) throw new TokenVerificationError("JWT does not contain all required scopes");
    }
    return { subject: claims.sub, issuer: this.issuer, audience: tokenAudiences, scopes, claims };
  }

  private fresh(): boolean {
    return this.keys.size > 0 && Date.now() - this.loadedAt < this.jwksTtlMs;
  }

  // throttled reports whether a refresh happened too recently to allow
  // another. A cached key is still returned by signingKey when it holds one,
  // so throttling never rejects a token the current document can verify.
  private throttled(): boolean {
    return this.lastRefresh > 0 && Date.now() - this.lastRefresh < this.jwksMinFetchMs;
  }

  private recentMiss(kid: string): boolean {
    const missedAt = this.unknownKids.get(kid);
    if (missedAt === undefined) return false;
    const now = Date.now();
    return now - missedAt < this.unknownKeyTtlMs && now - this.lastRefresh < this.jwksMinFetchMs;
  }

  private async signingKey(kid: string, signal?: AbortSignal): Promise<JsonWebKey | undefined> {
    const cached = this.keys.get(kid);
    if (cached && this.fresh()) return cached;
    if (this.recentMiss(kid)) return undefined;
    // The refresh interval is global, not per kid. Keying it on repeated
    // misses of the same kid let an unauthenticated caller send a stream of
    // unique kids and draw one JWKS request per token.
    if (this.throttled()) return cached;
    // A kid that is absent from a populated cache needs its own fetch. Joining
    // an ordinary load that is already in flight would observe a document that
    // was requested before this kid was known.
    const force = this.keys.size > 0 && !this.keys.has(kid);
    await this.load(signal, force);
    const found = this.keys.get(kid);
    if (found) this.unknownKids.delete(kid);
    else this.unknownKids.set(kid, Date.now());
    return found;
  }

  private async load(signal: AbortSignal | undefined, force: boolean): Promise<void> {
    if (this.loading) {
      const inflight = this.loading;
      await inflight;
      if (!force) return;
      if (this.loading) return this.loading;
    }
    if (this.loading) return this.loading;
    const { signal: requestSignal, release } = requestTimeout(this.timeoutMs, signal);
    let pending: Promise<void>;
    pending = this.fetchJwks(requestSignal).finally(() => {
      release();
      if (this.loading === pending) this.loading = undefined;
    });
    this.loading = pending;
    return pending;
  }


  private async fetchJwks(signal?: AbortSignal): Promise<void> {
    this.validateJwksUri();
    try {
      const init: RequestInit = { headers: { accept: "application/json" }, redirect: "error" };
      if (signal) init.signal = signal;
      const response = await this.fetcher(this.jwksUri, init);
      if (!response.ok) throw new Error(`JWKS returned HTTP ${response.status}`);
      const text = await response.text();
      if (Buffer.byteLength(text) > MAX_JWKS_BYTES) throw new Error("JWKS response is too large");
      const document: unknown = JSON.parse(text);
      if (!document || typeof document !== "object" || !Array.isArray((document as JWKSDocument).keys)) {
        throw new Error("JWKS response is invalid");
      }
      const next = new Map<string, JsonWebKey>();
      for (const key of (document as JWKSDocument).keys) {
        if (key.kty === "RSA" && typeof key.kid === "string") next.set(key.kid, key);
      }
      this.keys = next;
      const now = Date.now();
      this.loadedAt = now;
      this.lastRefresh = now;
    } catch (error) {
      throw new TokenVerificationError("unable to load JWKS", { cause: error });
    }
  }

  private validateJwksUri(): void {
    let url: URL;
    try { url = new URL(this.jwksUri); } catch { throw new TokenVerificationError("JWKS URI must be absolute"); }
    if (!['http:', 'https:'].includes(url.protocol)) throw new TokenVerificationError("JWKS URI must use HTTP(S)");
    if (!this.ssrfSafe) return;
    if (url.username || url.password) throw new TokenVerificationError("JWKS URI must not contain credentials");
    const loopback = ["localhost", "127.0.0.1", "[::1]"].includes(url.hostname);
    if (url.protocol !== "https:" && !loopback) throw new TokenVerificationError("JWKS URI must use HTTPS");
    if (isIP(url.hostname.replace(/^\[|\]$/g, "")) && !loopback) {
      throw new TokenVerificationError("JWKS URI must not use a non-loopback IP address");
    }
  }
}
